llama.cpp

```diff
@@ -13326,10 +13414,19 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
             qs.has_output = true;
         }
     }
-    if (qs.n_attention_wv != qs.n_ffn_down || (uint32_t) qs.n_attention_wv != model.hparams.n_layer) {
-        LLAMA_LOG_WARN("%s ============ Strange model: n_attention_wv = %d, n_ffn_down = %d, hparams.n_layer = %d\n",
-                __func__, qs.n_attention_wv, qs.n_ffn_down, model.hparams.n_layer);
-    }
+    // REVIEW: i do not undertand why there is logic for counting the number of layers by counting the number of tensors
+    // instead of just using the n_layer metadata
+    // without this change, it would require different logic for merged experts and split experts models,
+    // as split expert models end with a ffn_* count n_expert times higher than the real number of layers,
+    // which then is corrected in layer_info by dividing the value by n_expert
+    // this code needs to be refactored
```

Comment:
The effect is rather small - since `n_ffn_gate` was overestimated some of the last ffn tensors remained `IQ3_XXS` instead of the intended `IQ3_S` when the quantization mixture used is `IQ3_XS`:

https://github.com/ggerganov/llama.cpp/blob/52604860f93063ef98863921da697576af1c7665/llama.cpp#L13105-L13112

For example, Mixtral 8x7B:

- `master`

```
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:   32 tensors
llama_model_loader: - type q8_0:   64 tensors
llama_model_loader: - type q5_K:   32 tensors
llama_model_loader: - type q6_K:    1 tensors
llama_model_loader: - type iq3_xxs:  440 tensors
llama_model_loader: - type iq3_s:  361 tensors

llm_load_print_meta: model size       = 17.96 GiB (3.30 BPW) 
```

- PR

```
llama_model_loader: - type  f32:   65 tensors
llama_model_loader: - type  f16:   32 tensors
llama_model_loader: - type q8_0:   64 tensors
llama_model_loader: - type q5_K:   32 tensors
llama_model_loader: - type q6_K:    1 tensors
llama_model_loader: - type iq3_xxs:  416 tensors
llama_model_loader: - type iq3_s:  385 tensors

llm_load_print_meta: model size       = 18.02 GiB (3.31 BPW) 
```